KAFKA-4093 : Cluster Identifiers (KIP-78)

arrawatia opened the pull request:
This PR implements  KIP-78:Cluster Identifiers [(link)](https://cwiki.apache.org/confluence/display/KAFKA/KIP-78%3A+Cluster+Id#KIP-78:ClusterId-Overview) and includes the following changes:
1. Changes to broker code
   - generate cluster id and store it in Zookeeper
   - update protocol to add cluster id to metadata request and response
   - add ClusterResourceListener interface, ClusterResource class and ClusterMetadataListeners utility class
   - send ClusterResource events to the metric reporters
2. Changes to client code
   - update Cluster and Metadata code to support cluster id
   - update clients for sending ClusterResource events to interceptors, (de)serializers and metric reporters
3. Integration tests for interceptors, (de)serializers and metric reporters for clients and for protocol changes and metric reporters for broker.
4. System tests for upgrading from previous versions.

arrawatia committed:
First cut for cluster id.

1. Add v2 for Metadata Request and Response in Protocol to include
cluster_id
2. Updated the MetadataRequest and MetadataResponse to add cluster_id
and Update tests
3. Add cluster_id to KafkaServer and KafkaApis
4. cluster_id = dummy for now.

arrawatia committed:
Broker Id generation.

arrawatia committed:
Change ZK representation to JSON from string.

arrawatia committed:
Add tests and change from using internal sun classes to JAXB for base64 en(de)coding

arrawatia committed:
Make cluster_id nullable and change NO_CLUSTER_ID to null.

arrawatia committed:
First pass for client changes.

arrawatia committed:
Add cluster.id to meta.properties

arrawatia committed:
Change to ClusterListerners utility class.

arrawatia committed:
Merge remote-tracking branch 'confluentinc/trunk' into kip-78

arrawatia committed:
Revert change for adding cluster.id to meta.properties.

Wait for it to be included in the KIP.

arrawatia committed:
Make check style tests pass.

arrawatia committed:
Rename ClusterListener -> ClusterResourceListener and ClusterResourceMeta -> ClusterResource.

arrawatia committed:
Add ClusterListeners on KafkaServer to send events to MetricReporters(both Codahale and KafkaMetricrReporter).

arrawatia committed:
Update consumer tests to test for cluster id in the correct order.

arrawatia committed:
Make check style tests pass.

arrawatia committed:
Merge remote-tracking branch 'confluentinc/trunk' into kip-78

arrawatia committed:
Merge branch 'trunk' of github.com:apache/kafka into kip-78

arrawatia committed:
Fix case for ClusterIdPath constant.

arrawatia committed:
Add integration tests for cluster id generation.

arrawatia committed:
Make check style happy.

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListener.java:
```diff
@@ -0,0 +1,18 @@
+package org.apache.kafka.common;
+
+public interface ClusterResourceListener {
+    void onClusterUpdate(ClusterResource clusterMetadata);
```
The KIP was updated at some point to be just `onUpdate`.

junrao replied:
clusterMetadata probably should be clusterResource?

Also, onClusterUpdate() probably should be onClusterResourceUpdate()? Alternatively, would it be simpler to just name it onUpdate()?

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListeners.java:
```diff
@@ -0,0 +1,64 @@
+        return clusterResourceListeners;
+    }
+
+    public void add(Object candidate) {
```
Why is it `Object`?

arrawatia replied:
It is because this method needs to accept all types (MetricReporter, Interceptors, Serializers etc) that might or might not implement the ClusterListenerInterface.

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -35,6 +37,43 @@ class MetadataRequestTest extends BaseRequestTest {
+  }
+
+  @Test
+  def testClusterIdIsValidUUID() {
```
We generally use `Uuid` capitalisation in Kafka.

ijuma commented on core/src/test/scala/unit/kafka/server/ServerGenerateClusterIdTest.scala:
```diff
@@ -0,0 +1,113 @@
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+package unit.kafka.server
```
There should be no `unit` here.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -58,6 +60,7 @@
     /* Topics with expiry time */
     private final Map<String, Long> topics;
     private final List<Listener> listeners;
+    private ClusterResourceListeners clusterResourceListeners;
```
Should be final. Oh, we don't set it during the constructor. Would it be possible to set it in the constructor?

junrao replied:
Does this need to be volatile?

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListeners.java:
```diff
@@ -0,0 +1,64 @@
+        }
+    }
+
+    public static ClusterResourceListeners empty() {
```
Do we need this?

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListeners.java:
```diff
@@ -0,0 +1,64 @@
+        this.clusterResourceListeners = new ArrayList<>();
+    }
+
+    public List<ClusterResourceListener> getClusterResourceListeners() {
```
We generally don't use the `get` prefix in Kafka (yes, there are older bits of code that still do). Also, do we really want to expose the underlying list in a mutable form?

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -288,35 +308,47 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
+    val clusterId = zkUtils.getClusterId match {
+      case Some(id) => id
+      case None => {
+        val uuid:UUID = UUID.randomUUID()
```
No need to specify the type in cases like this.

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -421,7 +453,7 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
                 ioException = true
                 warn("Error during controlled shutdown, possibly because leader movement took longer than the configured socket.timeout.ms: %s".format(ioe.getMessage))
-                // ignore and try again
+              // ignore and try again
```
This change looks wrong. A few other examples. I'd try to avoid reformatting code outside of the code you are changing (unless the existing formatting is wrong).

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -35,6 +37,43 @@ class MetadataRequestTest extends BaseRequestTest {
+  def testClusterIdIsValidUUID() {
+    val metadataResponse = sendMetadataRequest(MetadataRequest.allTopics(), 2)
+    val clusterId = metadataResponse.clusterId()
+    val decodedUUID = DatatypeConverter.parseBase64Binary(clusterId)
```
Not sure if we want to use `DatatypeConverter`, is it supported API?

arrawatia replied:
Yes, this is part of the JDK[1]. 
The only other option in JDK7 is to use the internal `sun.misc.BASE64Encoder`/ `sun.misc.BASE64Decoder` classes. Otherwise, we will need to implement our own or add a third party dependency.

Note that once we upgrade to Java 8, we can use the new `Base64` class [2].
1. https://docs.oracle.com/javase/7/docs/api/javax/xml/bind/DatatypeConverter.html
2. http://docs.oracle.com/javase/8/docs/api/java/util/Base64.html

junrao commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -283,6 +290,10 @@ public synchronized void removeListener(Listener listener) {
         this.listeners.remove(listener);
     }
 
+    public void setClusterResourceListeners(ClusterResourceListeners clusterResourceListeners) {
```
Do we actually expect ClusterResourceListeners to be updated? if not, would it be better to pass in ClusterResourceListeners in the constructor?

junrao commented on clients/src/main/java/org/apache/kafka/common/Cluster.java:
```diff
@@ -250,6 +260,13 @@ public boolean isBootstrapConfigured() {
         return isBootstrapConfigured;
     }
 
+    public String clusterId() {
```
Do we need this method given getClusterResource()?

junrao commented on clients/src/main/java/org/apache/kafka/common/Cluster.java:
```diff
@@ -250,6 +260,13 @@ public boolean isBootstrapConfigured() {
+    }
+
+    public ClusterResource getClusterResource() {
+        return new ClusterResource(clusterId());
```
Could we just create a ClusterResource once on update and initialization?

ijuma replied:
Should be named `clusterResource()` without the `get`.

junrao commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListeners.java:
```diff
@@ -0,0 +1,64 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
+ * the License for the
+ * specific language governing permissions and limitations under the License.
+ */
```
The license header looks different from other files.

junrao commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListeners.java:
```diff
@@ -0,0 +1,64 @@
+
+        for (Object candidate : candidateList) {
+
+            if (candidate instanceof ClusterResourceListener) {
```
No need for {} for single line statement.

junrao commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -116,8 +118,17 @@
-    public static final Schema[] METADATA_RESPONSE = new Schema[] {METADATA_RESPONSE_V0, METADATA_RESPONSE_V1};
+    public static final Schema METADATA_RESPONSE_V2 = new Schema(new Field("brokers", new ArrayOf(METADATA_BROKER_V1),
+                                                                    "Host and port information for all brokers."),
+                                                                    new Field("cluster_id", NULLABLE_STRING,
```
indentation.

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -155,9 +167,9 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
-   */
+    * Start up API for bringing up a single instance of the Kafka server.
+    * Instantiates the LogManager, the SocketServer and the request handlers - KafkaRequestHandlers
+    */
```
It seems there is no need to remove the space in the above 3 lines.

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -217,14 +237,14 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
         /* start dynamic config manager */
         dynamicConfigHandlers = Map[String, ConfigHandler](ConfigType.Topic -> new TopicConfigHandler(logManager, config),
-                                                           ConfigType.Client -> new ClientIdConfigHandler(apis.quotaManagers))
+          ConfigType.Client -> new ClientIdConfigHandler(apis.quotaManagers))
```
The original indentation seems better? Ditto in a few places below.

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -494,7 +526,7 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
                 channel = null
                 warn("Error during controlled shutdown, possibly because leader movement took longer than the configured socket.timeout.ms: %s".format(ioe.getMessage))
-                // ignore and try again
+              // ignore and try again
```
The original indentation seems correct?

junrao commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -209,6 +210,43 @@ class ZkUtils(val zkClient: ZkClient,
+
+    try {
+      createParentPath(ClusterIdPath)
+      ZkPath.createPersistent(zkClient, ClusterIdPath, data, DefaultAcls)
```
Could we just call createPersistentPath() which already auto creates the parent?

arrawatia committed:
Integration tests for cluster id for all components that can implement ClusterResourceListerner interface.

arrawatia committed:
Rename file correctly.

arrawatia committed:
Fixes for review comments from @junrao and @ijuma.

arrawatia committed:
Merge remote-tracking branch 'confluentinc/trunk' into kip-78

arrawatia committed:
Fix license.

junrao commented on clients/src/main/java/org/apache/kafka/common/ClusterResource.java:
```diff
@@ -0,0 +1,39 @@
+        this.clusterId = clusterId;
+    }
+
+    public String getClusterId() {
```
Since we don't really use getters, it's better to use clusterId() instead of getClusterId().

junrao commented on clients/src/test/java/org/apache/kafka/test/MockConsumerInterceptor.java:
```diff
@@ -52,6 +59,11 @@ public void configure(Map<String, ?> configs) {
     @Override
     public ConsumerRecords<String, String> onConsume(ConsumerRecords<String, String> records) {
+        if (CLUSTER_META.get() != null
+                && CLUSTER_META.get().getClusterId() != null
```
indentation

junrao commented on core/src/main/scala/kafka/server/KafkaServerStartable.scala:
```diff
@@ -24,14 +24,19 @@ import kafka.utils.{VerifiableProperties, Logging}
 
+  // Add Coda Hale metric reporters which implement ClusterResourceListener interface to ClusterResourceListeners.
+  for(reporter <- reporters) {
+    server.clusterResourceListeners.add(reporter)
```
This seems a bit hacky. Would it be better to pass in the initial reporters as an optional field to the constructor of KafkaServer?

junrao commented on core/src/test/scala/integration/kafka/api/EndToEndClusterIdTest.scala:
```diff
@@ -0,0 +1,204 @@
+import scala.collection.JavaConverters._
+
+/** The test cases here verify the following:
+  * 1. The ProducerInterceptor receives the Cluster metadata after first send.
```
Is this accurate? The producer updates Metadata with the bootstrap cluster during initialization. So, it seems that ClusterResourceListener will be called before first send. It's just that the clusterId won't be set in the first call to the listener?

arrawatia replied:
You are correct. It will be available after the first send in general. What I meant to say was that the cluster id will be available to the interceptor after the first onSend() call on the interceptor. I will update the sentence to make that clear.

ijuma replied:
It seems a bit weird to call the listener before we have a cluster id since it's the only information we have at the moment. The KIP suggests that we will call the listener for the first time after we receive the cluster id.

arrawatia replied:
Ok, now that you mention it, I agree that it will be weird to the users who don't know how the client works internally. 

@ijuma Do you think I should only call the listener only if cluster id is present ? I can add that condition in the Metadata class. (I can't think of any reason the users will want to get a notification for the bootstrap cluster).

ijuma replied:
Yeah, the bootstrap cluster is an implementation detail in my opinion and there's no need to call the listener in that case. Adding the check in `Metadata` class seems like a reasonable first step. I'll take a closer look tomorrow and see if there's a better way, but feel free to update before then if you have the time (that way we'll have the right behaviour).

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -681,6 +689,9 @@ private KafkaConsumer(ConsumerConfig config,
 
+
+
+
```
Are these extra lines intentional?

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListeners.java:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+
+    public void addAll(List candidateList) {
```
We should not use raw types, it should be `List<?>` if the type cannot be checked.

arrawatia committed:
Changes for review comments from @junrao .

arrawatia committed:
Merge branch 'trunk' of github.com:apache/kafka into kip-78

arrawatia committed:
Fix error.

arrawatia committed:
Clarify the event order in the docs.

arrawatia committed:
Only send onUpdate events when cluster id is present.

arrawatia committed:
Add system test for testing that cluster id is set properly after upgrades.

arrawatia committed:
Add system test for testing upgrades from 0.10.x to trunk.

arrawatia committed:
Add Javadocs to ClusterResourceListener interface.

arrawatia committed:
Remove raw types as per comments from @ijuma.

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResource.java:
```diff
@@ -0,0 +1,37 @@
+
+package org.apache.kafka.common;
+
+public class ClusterResource {
```
We need javadoc for this class since it's a public class.

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResource.java:
```diff
@@ -0,0 +1,37 @@
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common;
```
I wonder if we should use a subpackage for this and `ClusterResourceListener` classes. What do you think @junrao and @hachikuji?

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListener.java:
```diff
@@ -0,0 +1,42 @@
+ * Cluster metadata.
+ * <p>
+ * Users who need access to cluster metadata in interceptors, metric reporters, serializers and deserializers
+ * can implement this interface. The order of method calls for each of these types is described below
```
We should add a brief pointer in those interfaces pointing to this one.

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListener.java:
```diff
@@ -0,0 +1,42 @@
+ * <p>
+ * {@link org.apache.kafka.common.metrics.MetricsReporter} : The {@link ClusterResourceListener#onUpdate(ClusterResource)} method will be called during the bootup of the Kafka broker. The reporter may receive metric events from the network layer before this method is called.
+ * <p>
+ * KafkaMetricsReporter : The {@link ClusterResourceListener#onUpdate(ClusterResource)} method will be called during the bootup of the Kafka broker. The reporter may receive metric events from the network layer before this method is called.
```
It's worth having a note about the fact that the cluster id may be null when talking to brokers before 0.10.1.0. If an implementor of this listener receives a `null` cluster id, they can expect it to always be `null` unless they have a cluster with mixed broker versions or the cluster is upgraded while the client is running.

ijuma replied:
Also maybe we can structure this description so that there's broker and client sections. For the broker, there is a single invocation on start-up and the id should never change. For the clients, there is one invocation after each metadata response and it is possible for the cluster id to go from null to non-null as per the comment above.

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListeners.java:
```diff
@@ -0,0 +1,51 @@
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common;
```
This is an internal class and should live in an `internals` package.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -44,6 +44,9 @@
     public static final int NO_CONTROLLER_ID = -1;
 
+    private static final String CLUSTER_ID_KEY_NAME = "cluster_id";
+    public static final String NO_CLUSTER_ID = null;
```
We should probably not bother with this. We use constants like this when using magic values like `-1`. For nullable values, it's fine to just use plain `null`. Particularly as this constant lives in an internal class, so users of `ClusterResource` can't use it anyway.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -79,19 +82,19 @@
     private final Node controller;
     private final List<TopicMetadata> topicMetadata;
 
+
```
Unintentional line?

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -295,6 +302,14 @@ public Node controller() {
+        } else {
+            return NO_CLUSTER_ID;
+        }
+    }
```
Why are we doing it this way instead of following the same pattern as for controller id where we set the field in the constructor?

ijuma commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -145,6 +147,28 @@ public void testUpdateWithNeedMetadataForAllTopics() {
+        ClusterResourceListeners listeners = new ClusterResourceListeners();
+        listeners.add(mockClusterListener);
+        metadata = new Metadata(refreshBackoffMs, metadataExpireMs, false, listeners);
+        metadata.update(Cluster.empty(), time);
```
We should check that this first `update` doesn't trigger a notification.

ijuma commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -289,6 +289,9 @@ public void testInterceptorConstructorClose() throws Exception {
             assertEquals(1, MockConsumerInterceptor.INIT_COUNT.get());
             assertEquals(1, MockConsumerInterceptor.CLOSE_COUNT.get());
+            // Cluster metadata will only be updated on calling poll.
+            Assert.assertNull(MockConsumerInterceptor.CLUSTER_META.get());
```
Is there a test in this class where we can do a non-null assertion?

arrawatia replied:
No. I have those tests in the integration tests. Doing a non-null test here will need a lot of mocking at the n/w layer. Do you think it adds value if it is already being tested in the integration tests ?

ijuma replied:
No, let's leave it as is.

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:
```diff
@@ -104,6 +104,9 @@ public void testInterceptorConstructClose() throws Exception {
             Assert.assertEquals(0, MockProducerInterceptor.CLOSE_COUNT.get());
 
+            // Cluster metadata will only be updated on calling onSend.
+            Assert.assertNull(MockProducerInterceptor.CLUSTER_META.get());
```
Is there a test in this class where we can do a non-null assertion?

ijuma commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -73,8 +73,8 @@ public void testSerialization() throws Exception {
-                createMetadataRequest(Arrays.asList("topic1")).getErrorResponse(1, new UnknownServerException()),
-                createMetadataResponse(1),
+                createMetadataRequest(Arrays.asList("topic1")).getErrorResponse(2, new UnknownServerException()),
+                createMetadataResponse(2),
```
We should also test version 1. A few lines below we have:

``` java
createMetadataResponse(0);
createMetadataRequest(Arrays.asList("topic1")).getErrorResponse(0, new UnknownServerException());
```

You can add a couple more lines there.

ijuma commented on clients/src/test/java/org/apache/kafka/test/MockConsumerInterceptor.java:
```diff
@@ -52,6 +59,9 @@ public void configure(Map<String, ?> configs) {
     @Override
     public ConsumerRecords<String, String> onConsume(ConsumerRecords<String, String> records) {
+        if (CLUSTER_META.get() != null && CLUSTER_META.get().clusterId() != null && CLUSTER_META.get().clusterId().length() == 48)
+            IS_CLUSTER_ID_PRESENT_BEFORE_ON_CONSUME.set(true);
```
Instead of doing this, you can just set the cluster id on another variable called `CLUSTER_META_BEFORE_ON_CONSUME` via a compare and set (so that it's only done once).

ijuma replied:
That way logic like the length, what the field is, etc. can live outside this class that should be agnostic.

ijuma commented on clients/src/test/java/org/apache/kafka/test/MockDeserializer.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
```
One extra line.

ijuma commented on clients/src/test/java/org/apache/kafka/test/MockDeserializer.java:
```diff
@@ -0,0 +1,60 @@
+    @Override
+    public byte[] deserialize(String topic, byte[] data) {
+        if (CLUSTER_META.get() != null && CLUSTER_META.get().clusterId() != null && CLUSTER_META.get().clusterId().length() == 48)
+            IS_CLUSTER_ID_PRESENT_BEFORE_DESERIALIZE.set(true);
```
Same comment as in `MockConsumerInterceptor`.

ijuma commented on clients/src/test/java/org/apache/kafka/test/MockMetricsReporter.java:
```diff
@@ -51,4 +51,5 @@ public void close() {
     public void configure(Map<String, ?> configs) {
 
     }
+
```
Unintentional?

ijuma commented on clients/src/test/java/org/apache/kafka/test/MockProducerInterceptor.java:
```diff
@@ -20,19 +20,25 @@
     public static final AtomicInteger ON_ERROR_WITH_METADATA_COUNT = new AtomicInteger(0);
+    public static final AtomicReference<ClusterResource> CLUSTER_META = new AtomicReference<>();
+
+
```
One extra line.

ijuma commented on clients/src/test/java/org/apache/kafka/test/MockProducerInterceptor.java:
```diff
@@ -86,5 +92,11 @@ public static void resetCounters() {
+
+    @Override
+    public void onUpdate(ClusterResource clusterResource) {
+        CLUSTER_META.set(clusterResource);
```
Do you also want to verify that the cluster meta is set before `onAcknowledgement`?

ijuma commented on clients/src/test/java/org/apache/kafka/test/MockSerializer.java:
```diff
@@ -16,14 +16,22 @@
+    public static final AtomicReference<ClusterResource> CLUSTER_META = new AtomicReference<>();
+    public static final AtomicBoolean IS_CLUSTER_ID_PRESENT_BEFORE_SERIALIZE = new AtomicBoolean();
+
+
```
Extra lines?

ijuma commented on clients/src/test/java/org/apache/kafka/test/MockSerializer.java:
```diff
@@ -35,11 +43,18 @@ public void configure(Map<String, ?> configs, boolean isKey) {
     @Override
     public byte[] serialize(String topic, byte[] data) {
+        if (CLUSTER_META.get() != null && CLUSTER_META.get().clusterId() != null && CLUSTER_META.get().clusterId().length() == 48)
+            IS_CLUSTER_ID_PRESENT_BEFORE_SERIALIZE.set(true);
```
Same comment as for other similar code.

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -138,6 +141,8 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
   val brokerMetaPropsFile = "meta.properties"
   val brokerMetadataCheckpoints = config.logDirs.map(logDir => (logDir, new BrokerMetadataCheckpoint(new File(logDir + File.separator +brokerMetaPropsFile)))).toMap
+  var clusterId: String = null
+  val clusterResourceListeners  = new ClusterResourceListeners()
```
The new fields should be private. Maybe take the chance and change the existing ones to be private too.

arrawatia replied:
I tried to make the existing fields private and a whole lot of tests break. I also need access to cluster id in my tests. What is best way to make them private ? Add a getter ?

ijuma replied:
Let's leave it as it is for this PR then.

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -181,6 +193,19 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
+        }
+        // Add reporters for Kafka metrics framework (MetricsReporter) who implement ClusterResourceListener
+        clusterResourceListeners.addAll(reporters)
+        clusterResourceListeners.onUpdate(new ClusterResource(clusterId))
```
Can we extract the listener logic to a separate method please?

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -305,6 +330,18 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
+
+    val clusterId = zkUtils.getClusterId match {
+      case Some(id) => id
+      case None => {
```
Brace is not needed.

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -305,6 +330,18 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
+        zkUtils.createOrGetClusterId(base64EncodedUUID)
+      }
+    }
+    clusterId
```
This can be written as:

``` scala
zkUtils.getClusterId.getOrElse {
  val uuid = ...
  ....
}
```

ijuma commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -209,6 +210,42 @@ class ZkUtils(val zkClient: ZkClient,
+    readDataMaybeNull(ClusterIdPath)._1 match {
+      case Some(clusterId) => Some(parseClusterId(clusterId))
+      case None => None
+    }
```
Use `map` instead of pattern matching here.

ijuma commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -209,6 +210,42 @@ class ZkUtils(val zkClient: ZkClient,
+          val clusterIdMap = m.asInstanceOf[Map[String, Any]]
+          clusterIdMap.get("id").get.asInstanceOf[String]
+        case None => throw new KafkaException("Failed to parse the cluster id json [%s].".format(clusterIdString))
+      }
```
You can use `map/getOrElse` instead of this. Something like:

``` scala
Json.parseFull(clusterIdString).map { m =>
  ...
}.getOrElse(throw new KafkaException(...))
```

Also, use string interpolation (`s"$foo"`) instead of `format`.

ijuma commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -706,6 +712,7 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     testConsumer.assign(List(tp).asJava)
     testConsumer.seek(tp, 0)
 
+
```
Unintentional new line?

ijuma commented on tests/kafkatest/tests/core/upgrade_test.py:
```diff
@@ -112,3 +114,29 @@ def test_upgrade(self, from_kafka_version, to_message_format_version, compressio
+
+    @parametrize(from_kafka_version=str(LATEST_0_9))
+    @parametrize(from_kafka_version=str(LATEST_0_10))
+    def test_upgrade_for_cluster_id(self, from_kafka_version):
```
Why do we need another test for this? Can we not add the asserts in the existing test?

ijuma replied:
The reason is that it would be faster than spinning up a whole cluster just for the cluster id. Also, we would want to test more versions. I suggest adding `LATEST_0_10` to `test_upgrade`. What do you think?

ijuma commented on tests/kafkatest/version.py:
```diff
@@ -77,4 +77,5 @@ def get_version(node=None):
 V_0_10_0_0 = KafkaVersion("0.10.0.0")
-LATEST_0_10 = V_0_10_0_0
+V_0_10_0_1 = KafkaVersion("0.10.0.1")
+LATEST_0_10 = V_0_10_0_1
```
I think we may want a LATEST_0_10_0 because the next release will be 0.10.1.0 and would qualify as LATEST_0_10

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -60,6 +60,8 @@
                                                                           ArrayOf.nullable(STRING),
                                                                           "An array of topics to fetch metadata for. If the topics array is null fetch metadata for all topics."));
 
+    public static final Schema METADATA_REQUEST_V2 = METADATA_REQUEST_V1;
```
Please add a comment explaining that for the request v1 and v2 are the same, but the response for v2 is different.

ijuma commented on core/src/test/scala/unit/kafka/server/ServerGenerateClusterIdTest.scala:
```diff
@@ -0,0 +1,113 @@
+import org.junit.Assert._
+import org.junit.{Before, Test}
+
+class ServerGenerateClusterIdTest extends ZooKeeperTestHarness {
```
Isn't it better to inherit from `KafkaServerTestHarness`?

ijuma replied:
OK, I guess you want to be able to create the server yourself.

ijuma commented on core/src/test/scala/unit/kafka/server/ServerGenerateClusterIdTest.scala:
```diff
@@ -0,0 +1,113 @@
+
+    // Make sure the cluster id is 48 characters long (base64 of UUID.randomUUID() )
+    val clusterIdOnFirstBoot = server1.clusterId
+    assertEquals(clusterIdOnFirstBoot.length, 48)
```
We should probably verify that it only has valid characters.

ijuma replied:
Also that it's a valid UUID as you did in the metadata test.

ijuma commented on core/src/test/scala/unit/kafka/server/ServerGenerateClusterIdTest.scala:
```diff
@@ -0,0 +1,113 @@
+
+    //Make sure that the cluster id is persistent.
+    assertTrue(zkUtils.pathExists(ZkUtils.ClusterIdPath))
+    assertEquals(zkUtils.getClusterId().get, clusterIdOnFirstBoot)
```
Avoid the `get` and add `Some` to the right value. That way you get a more useful error message if the test fails.

ijuma commented on core/src/test/scala/unit/kafka/server/ServerGenerateClusterIdTest.scala:
```diff
@@ -0,0 +1,113 @@
+
+    //Make sure that the cluster id is persistent after multiple reboots.
+    assertTrue(zkUtils.pathExists(ZkUtils.ClusterIdPath))
+    assertEquals(zkUtils.getClusterId().get, clusterIdOnFirstBoot)
```
Same as above.

ijuma commented on core/src/test/scala/unit/kafka/server/ServerGenerateClusterIdTest.scala:
```diff
@@ -0,0 +1,113 @@
+    //Make sure that the cluster id doesn't exist yet.
+    assertFalse(zkUtils.pathExists(ZkUtils.ClusterIdPath))
+    var server1 = new KafkaServer(config1, threadNamePrefix = Option(this.getClass.getName))
+    server1.startup()
```
You can use `TestUtils.createServer` to replace the two lines above (there are a few other cases like that.

ijuma commented on core/src/test/scala/unit/kafka/server/ServerGenerateClusterIdTest.scala:
```diff
@@ -0,0 +1,113 @@
+    props2 = TestUtils.createBrokerConfig(2, zkConnect)
+    config2 = KafkaConfig.fromProps(props2)
+    props3 = TestUtils.createBrokerConfig(3, zkConnect)
+    config3 = KafkaConfig.fromProps(props3)
```
I think you can do all of this when the var (which would become val) is initialised. It doesn't involve any real resources right?

ijuma commented on core/src/test/scala/unit/kafka/server/ServerGenerateClusterIdTest.scala:
```diff
@@ -0,0 +1,113 @@
+    server2.startup()
+    val clusterIdFromServer2 = server2.clusterId
+    server3.startup()
+    val clusterIdFromServer3 = server3.clusterId
```
We should probably start all servers in parallel. You can do it by using scala's `Future`. Something like:

``` scala
import scala.concurrent._
import ExecutionContext.Implicits._
import duration.Duration

val future = Future.traverse(Seq(config1, config2, config3))(TestUtils.createServer(_))
val Seq(server1, server2, server3) = Await.result(future, Duration.Inf) // replace `Inf` with an appropriate time to wait
```

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -35,6 +37,43 @@ class MetadataRequestTest extends BaseRequestTest {
+  }
+
+  @Test
+  def testClusterIdWithNoTopics() {
```
I think this should be named `testClusterIdWithRequestVersion1` and we don't need `testClusterIdWithAllTopics`.

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -35,6 +37,43 @@ class MetadataRequestTest extends BaseRequestTest {
+
+    val v1MetadataResponse = sendMetadataRequest(MetadataRequest.allTopics(), 1)
+    val v1ClusterId = v1MetadataResponse.clusterId()
+    assertEquals(s"v1 clusterId should be ${MetadataResponse.NO_CLUSTER_ID}", v1ClusterId, MetadataResponse.NO_CLUSTER_ID)
```
`assertNull` seems fine here.

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -35,6 +37,43 @@ class MetadataRequestTest extends BaseRequestTest {
+  def testClusterIdWithNoTopics() {
+    val v2MetadataResponse = sendMetadataRequest(new MetadataRequest(List[String]().asJava), 2)
+    val v2ClusterId = v2MetadataResponse.clusterId()
+    assertNotNull("clusterId should not be null", v2ClusterId)
```
We probably don't need the v2 request in this test.

ijuma commented on core/src/test/scala/unit/kafka/server/KafkaMetricReporterClusterIdTest.scala:
```diff
@@ -0,0 +1,69 @@
+}
+
+object MockKafkaMetricsReporter {
+  val CLUSTER_META: AtomicReference[ClusterResource] = new AtomicReference[ClusterResource]
```
No need to repeat `AtomicReference` on both sides, the type annotation on the left can be removed.

ijuma commented on core/src/test/scala/unit/kafka/server/KafkaMetricReporterClusterIdTest.scala:
```diff
@@ -0,0 +1,69 @@
+
+    // Make sure the cluster id is 48 characters long (base64 of UUID.randomUUID() )
+    assertNotNull(MockKafkaMetricsReporter.CLUSTER_META)
+    assertEquals(48, MockKafkaMetricsReporter.CLUSTER_META.get().clusterId().length())
```
This is not right, the cluster id should be 22 characters long.

ijuma replied:
I suggest extracting a utility method that checks the validity, length and characters of the UUID and reuse it everywhere.

ijuma commented on core/src/test/scala/unit/kafka/server/KafkaMetricReporterClusterIdTest.scala:
```diff
@@ -0,0 +1,69 @@
+    server1.shutdown()
+
+    CoreUtils.delete(config1.logDirs)
+    TestUtils.verifyNonDaemonThreadsStatus(this.getClass.getName)
```
We are leaking servers and such in some of these tests in case of failures. I suggest adding a createServer method that stores all the instantiated servers and does the shutdown and delete on `tearDown`. You may be able to use `KafkaServerTestHarness` with an implementation of `generateConfigs` that returns an empty `Seq`.

ijuma commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -684,7 +685,12 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    // Send one record and make sure clusterId is set after send
+    testProducer.send(new ProducerRecord(tp.topic(), tp.partition(), s"key 0", s"value 0"))
+    assertNotNull(MockProducerInterceptor.CLUSTER_META)
+    assertEquals(48, MockProducerInterceptor.CLUSTER_META.get().clusterId().length())
```
This seems brittle as the `send` is asynchronous.

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndClusterIdTest.scala:
```diff
@@ -0,0 +1,204 @@
+  val CLUSTER_META: AtomicReference[ClusterResource] = new AtomicReference[ClusterResource]
+}
+
+class MockBrokerMetricsReporter extends MockMetricsReporter with ClusterResourceListener {
```
If you have classes that are only used in a particular test, they should live inside a companion object, e.g.

``` scala
object EndToEndClusterIdTest {
// classes go here.
}
```

If they are used by multiple tests, they should live in their own files.

There's a few other tests that have the same issue.

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndClusterIdTest.scala:
```diff
@@ -0,0 +1,204 @@
+    assertEquals(48, MockBrokerMetricsReporter.CLUSTER_META.get().clusterId().length())
+
+
+
```
Extra lines?

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndClusterIdTest.scala:
```diff
@@ -0,0 +1,204 @@
+import scala.collection.JavaConverters._
+
+/** The test cases here verify the following conditions.
+  * 1. The ProducerInterceptor receives the cluster id after the onSend() method is called.
```
And before `onAckowledge`?

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndClusterIdTest.scala:
```diff
@@ -0,0 +1,204 @@
+  * 4. The ConsumerInterceptor receives the cluster id before the onConsume() method.
+  * 5. The Deserializer receives the cluster id before the deserialize() method is called.
+  * 6. The consumer MetricReporter receives the cluster id  after poll() is called on KafkaConsumer.
+  * 7. The broker MetricReporter receives the cluster id  after the broker startup is over.
```
Multiple spaces after `id`. Other lines may have the same issue.

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndClusterIdTest.scala:
```diff
@@ -0,0 +1,204 @@
+  * 9. All the components receive the same cluster id.
+  */
+
+
```
Multiple empty lines?

ijuma commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -209,6 +210,42 @@ class ZkUtils(val zkClient: ZkClient,
+  def createOrGetClusterId(proposedClusterId: String): String = {
+    val jsonMap = Map("version" -> "1",
+      "id" -> proposedClusterId
+    )
```
Maybe we can have an inner class that has both `toJson` and `fromJson` methods so that the logic for this is together. Also, it's worth adding a comment for the JSON format and do a simple roundtrip test.

ijuma commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -209,6 +210,42 @@ class ZkUtils(val zkClient: ZkClient,
+        getClusterId match {
+          case Some(clusterId) => clusterId
+          case None => throw new KafkaException("Failed to get cluster id from Zookeeper. This can only happen if /cluster/id is deleted from Zookeeper.")
+        }
```
Like elsewhere, you can do `getOrElse` instead of pattern matching.

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -181,6 +193,19 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
+
+        /* Send events to metric reporters who implement ClusterResourceListener */
+        // Add reporters for Codahale metrics framework (KafkaMetricsReporter) who implement ClusterResourceListener
+        for(kafkaMetricsReporter <- kafkaMetricsReporters) {
```
Space after `for`.

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -181,6 +193,19 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
+        /* Send events to metric reporters who implement ClusterResourceListener */
+        // Add reporters for Codahale metrics framework (KafkaMetricsReporter) who implement ClusterResourceListener
+        for(kafkaMetricsReporter <- kafkaMetricsReporters) {
+          clusterResourceListeners.add(kafkaMetricsReporter)
```
Why can't you use `addAll` for this case too?

ijuma commented on core/src/main/scala/kafka/metrics/KafkaMetricsReporter.scala:
```diff
@@ -50,14 +50,16 @@ trait KafkaMetricsReporter {
   val ReporterStarted: AtomicBoolean = new AtomicBoolean(false)
 
-  def startReporters (verifiableProps: VerifiableProperties) {
+  def startReporters (verifiableProps: VerifiableProperties):List[KafkaMetricsReporter] = {
```
Space after `:`

ijuma commented on core/src/main/scala/kafka/metrics/KafkaMetricsReporter.scala:
```diff
@@ -50,14 +50,16 @@ trait KafkaMetricsReporter {
           metricsConfig.reporters.foreach(reporterType => {
             val reporter = CoreUtils.createObject[KafkaMetricsReporter](reporterType)
             reporter.init(verifiableProps)
+            reporters ++= List(reporter)
```
Appending to `scala.List` is slow (prepending is fast). You should probably just use an `ArrayBuffer` and return `Seq` from the method definition.

ijuma commented on core/src/main/scala/kafka/metrics/KafkaMetricsReporter.scala:
```diff
@@ -50,14 +50,16 @@ trait KafkaMetricsReporter {
 
-  def startReporters (verifiableProps: VerifiableProperties) {
+  def startReporters (verifiableProps: VerifiableProperties):List[KafkaMetricsReporter] = {
+    var reporters = List[KafkaMetricsReporter]()
```
This should probably be inside the `synchronized` block. I think we should probably return `Option[Seq[KafkaMetricsReporter]]`. `None` would be returned if the start had already been done. An alternative is to store the reporters in a field and return the stored ones in that case.

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -147,6 +152,13 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
   newGauge(
+    "ClusterId",
+    new Gauge[String] {
+      def value = clusterId
```
Do we have a test that checks this?

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListeners.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+
+    public void add(Object candidate) {
```
I think we should call this `maybeAdd` as it will only add if the type is correct. It may make sense to have a normal `add` method that takes an instance of `ClusterResourceListener`.

ijuma replied:
And probably worth adding a comment explaining why we need `maybeAdd`.

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListeners.java:
```diff
@@ -0,0 +1,51 @@
+        this.clusterResourceListeners = new ArrayList<>();
+    }
+
+
```
Extra line.

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResourceListeners.java:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+
+    public void addAll(List<?> candidateList) {
```
Same as for `add`. We may want two variants of this method with appropriate names.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -607,7 +608,36 @@ private KafkaConsumer(ConsumerConfig config,
+            clusterResourceListeners.addAll(interceptorList);
+            clusterResourceListeners.addAll(reporters);
+            clusterResourceListeners.add(keyDeserializer);
+            clusterResourceListeners.add(valueDeserializer);
```
Can we extract these 5 lines into a separate method?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -224,7 +225,35 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
+            clusterResourceListeners.addAll(interceptorList);
+            clusterResourceListeners.addAll(reporters);
+            clusterResourceListeners.add(keySerializer);
+            clusterResourceListeners.add(valueSerializer);
```
It would be good to extract these 5 lines into a separate method.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -221,6 +229,11 @@ else if (expireMs <= now) {
 
+        // Only send events when cluster id is present.
+        if (cluster != null && cluster.clusterResource() != null && cluster.clusterResource().clusterId() != null) {
+            clusterResourceListeners.onUpdate(cluster.clusterResource());
```
The KIP says we would log the cluster id whenever it changed.

ijuma committed:
Remove a couple of `Metadata` and `Cluster` constructors, tweak listener notification in `Metadata` and a few other tweaks

Some notes: Cluster is a public class, but I removed a constructor that was introduced in trunk only. Removing that constructor
uncovered a bug in `Cluster.withPartitions`. Metadata is not a public class and we can change it as we want.

ijuma commented:
Thanks for the PR @arrawatia, I did an in-depth review and left a few comments and a PR with some improvements (https://github.com/arrawatia/kafka/pull/1). I think they should be easy to address. Functionally, the important thing is to figure out why the id is 48 characters in the existing code. I expected it to be much less (22).

arrawatia committed:
Merge pull request #1 from ijuma/kafka-4093-cluster-id

Remove a couple of `Metadata` and `Cluster` constructors, tweak listener notification in `Metadata` and a few other tweaks

arrawatia committed:
Fixes for @ijuma 's review comments.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerInterceptor.java:
```diff
@@ -32,6 +32,8 @@
  * <p>
  * ConsumerInterceptor callbacks are called from the same thread that invokes {@link org.apache.kafka.clients.consumer.KafkaConsumer#poll(long)}.
+ * <p>
+ * If you need access to cluster metadata, you will need to implement {@link org.apache.kafka.common.ClusterResourceListener} interface.
```
Nitpick: this would read a bit better like (same for other classes that have the same blurb):

"Implement {@link org.apache.kafka.common.ClusterResourceListener} to receive cluster metadata once it's available. Please see the class documentation for ClusterResourceListener for more information."

ijuma commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -308,6 +326,8 @@ private Cluster getClusterForCurrentTopics(Cluster cluster) {
         }
-        return new Cluster(nodes, partitionInfos, unauthorizedTopics, internalTopics);
+
+
```
Extra line.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1413,6 +1416,16 @@ public void wakeup() {
+        ClusterResourceListeners clusterResourceListeners = new ClusterResourceListeners();
+        for (List<?> candidateList: candidateLists) {
+            clusterResourceListeners.maybeAddAll(candidateList);
+        }
```
Nitpick: no need for braces.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -272,6 +297,8 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
                     metrics,
                     time);
+
+
```
Extra line.

ijuma commented on clients/src/main/java/org/apache/kafka/common/ClusterResource.java:
```diff
@@ -0,0 +1,44 @@
+
+    private final String clusterId;
+
+    public ClusterResource(String clusterId) {
```
It would be good to mention that clusterId can be null in the constructor.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -78,23 +80,24 @@
     private final Collection<Node> brokers;
     private final Node controller;
     private final List<TopicMetadata> topicMetadata;
+    private String clusterId = null;
```
This should be final. You need an `if/else` in the constructor and assign it to null in case it's not available.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -268,7 +279,7 @@ public Cluster cluster() {
         }
 
-        return new Cluster(this.brokers, partitions, topicsByError(Errors.TOPIC_AUTHORIZATION_FAILED), internalTopics);
+        return new Cluster(this.clusterId(), this.brokers, partitions, topicsByError(Errors.TOPIC_AUTHORIZATION_FAILED), internalTopics);
```
You can use the field `clusterId` instead of the method.

ijuma commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -113,7 +113,9 @@ public void testSerialization() throws Exception {
         createMetadataResponse(0);
+        createMetadataResponse(1);
         createMetadataRequest(Arrays.asList("topic1")).getErrorResponse(0, new UnknownServerException());
+        createMetadataRequest(Arrays.asList("topic1")).getErrorResponse(1, new UnknownServerException());
```
I just noticed that there's an existing bug here, we should pass these to `checkSerialization`.

ijuma commented on clients/src/test/java/org/apache/kafka/test/MockConsumerInterceptor.java:
```diff
@@ -32,11 +33,15 @@
     public static final AtomicInteger CLOSE_COUNT = new AtomicInteger(0);
     public static final AtomicInteger ON_COMMIT_COUNT = new AtomicInteger(0);
+    public static final AtomicReference<ClusterResource> CLUSTER_META = new AtomicReference<>();
+    public static final ClusterResource NO_CLUSTER_ID = new ClusterResource("foo");
```
Maybe instead of `foo`, you can say "no cluster id"

ijuma commented on clients/src/test/java/org/apache/kafka/test/MockDeserializer.java:
```diff
@@ -0,0 +1,59 @@
+    public static final AtomicInteger INIT_COUNT = new AtomicInteger(0);
+    public static final AtomicInteger CLOSE_COUNT = new AtomicInteger(0);
+    public static final AtomicReference<ClusterResource> CLUSTER_META = new AtomicReference<>();
+    public static final Object NO_CLUSTER_ID = new Object();
```
Maybe the approach you used in `MockConsumerInterceptor` is better as you can use the right type in `AtomicReference`.

ijuma commented on clients/src/test/java/org/apache/kafka/test/TestUtils.java:
```diff
@@ -269,4 +274,29 @@ public static void waitForCondition(final TestCondition testCondition, final lon
+     * @param clusterId
+     * @return true if valid, false otherwise.
+     */
+    public static boolean isValidClusterId(String clusterId) {
```
Maybe this method can use JUnit assertions for better error messages (instead of returning `boolean`).

ijuma commented on core/src/main/scala/kafka/metrics/KafkaMetricsReporter.scala:
```diff
@@ -42,22 +44,27 @@ trait KafkaMetricsReporterMBean {
+  var reporters: ArrayBuffer[KafkaMetricsReporter] = null
 
-  def startReporters (verifiableProps: VerifiableProperties) {
+  def startReporters (verifiableProps: VerifiableProperties): List[KafkaMetricsReporter] = {
```
We should just return `Seq` and then you don't need the conversion at the end

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -181,6 +194,13 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
+        clusterId = getOrGenerateClusterId(zkUtils)
+        info(s"Cluster ID = $clusterId")
+
+        /* Send events to metric reporters who implement ClusterResourceListener */
```
The comment seems redundant since the method name is clear.

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -271,6 +291,14 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
+    val clusterResourceListeners = new ClusterResourceListeners()
+    for (clusterListener <- clusterListeners) {
+      clusterResourceListeners.maybeAdd(clusterListener)
+    }
```
No brace needed.

ijuma commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -209,6 +210,36 @@ class ZkUtils(val zkClient: ZkClient,
   }
 
+  /* Represents a cluster identifier. Stored in Zookeeper in JSON format: {"version" -> "1", "id" -> id } */
+  object ClusterIdentifier {
```
Should probably just call it `ClusterId` as that is the name we've been using for some time.

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -35,6 +35,26 @@ class MetadataRequestTest extends BaseRequestTest {
   }
 
   @Test
+  def testClusterIdWithRequestVersion2() {
```
Does this test add any value over `testClusterIdIsValid`? It doesn't seem to me that it does.

ijuma commented on tests/kafkatest/tests/core/upgrade_test.py:
```diff
@@ -60,6 +62,7 @@ def perform_upgrade(self, from_kafka_version, to_message_format_version=None):
             self.kafka.start_node(node)
 
 
+    @parametrize(from_kafka_version=str(LATEST_0_10), to_message_format_version=None, compression_types=["none"])
```
We should use LATEST_0_10_0 here.

ijuma commented on core/src/test/scala/unit/kafka/server/ServerGenerateClusterIdTest.scala:
```diff
@@ -0,0 +1,144 @@
+    assertEquals(clusterIdFromServer1, server1.clusterId)
+    server2.startup()
+    assertEquals(clusterIdFromServer2, server2.clusterId)
+    server3.startup()
```
We should probably start them in parallel again.

ijuma reviewed (changes requested):
Thanks for the updates Sumit, left a few more comments.

arrawatia committed:
Make cluster id 24 characters long.

arrawatia committed:
Fixes for review comments by @ijuma.

ijuma commented on clients/src/test/java/org/apache/kafka/test/TestUtils.java:
```diff
@@ -269,4 +277,34 @@ public static void waitForCondition(final TestCondition testCondition, final lon
+    /**
+     * Checks if a cluster id is valid.
+     * @param clusterId
+     * @return true if valid, false otherwise.
```
This is no longer throw.

ijuma commented on clients/src/test/java/org/apache/kafka/test/TestUtils.java:
```diff
@@ -269,4 +277,34 @@ public static void waitForCondition(final TestCondition testCondition, final lon
+        //Check if it can be converted back to a UUID.
+        try {
+            ByteBuffer uuidBuffer = ByteBuffer.wrap(decodedUuid);
+            UUID.fromString(new UUID(uuidBuffer.getLong(), uuidBuffer.getLong()).toString());
```
Is there a case where `UUID.fromString` fails when given the output of `UUID.toString`? If not, maybe it's redundant.

ijuma commented on core/src/main/scala/kafka/metrics/KafkaMetricsReporter.scala:
```diff
@@ -42,22 +44,27 @@ trait KafkaMetricsReporterMBean {
 
 object KafkaMetricsReporter {
   val ReporterStarted: AtomicBoolean = new AtomicBoolean(false)
+  var reporters: ArrayBuffer[KafkaMetricsReporter] = null
```
This should be private as it's unsafe to access without synchronization.

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -138,6 +143,7 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
   val correlationId: AtomicInteger = new AtomicInteger(0)
   val brokerMetaPropsFile = "meta.properties"
   val brokerMetadataCheckpoints = config.logDirs.map(logDir => (logDir, new BrokerMetadataCheckpoint(new File(logDir + File.separator +brokerMetaPropsFile)))).toMap
+  var clusterId: String = null
```
So, I think it's fine not to change other fields to private, but we should make this one private. It's easy to provide an accessor via a `def`. The reason to make it private is that it should not be available for mutation outside this class. If we need an accessor, we can name it `clusterId` and rename the variable `_clusterId` or something like that.

ijuma commented on core/src/main/scala/kafka/utils/CoreUtils.scala:
```diff
@@ -277,4 +279,18 @@ object CoreUtils extends Logging {
     listenerList.map(listener => EndPoint.createEndPoint(listener)).map(ep => ep.protocolType -> ep).toMap
   }
+
+  def generateUuidAsBase64():String = {
```
Space after `:`

ijuma commented on core/src/main/scala/kafka/utils/CoreUtils.scala:
```diff
@@ -277,4 +279,18 @@ object CoreUtils extends Logging {
+  def generateUuidAsBase64():String = {
+    val uuid = UUID.randomUUID()
+    // Extract bytes for uuid which is 128 bits (or 16 bytes) long.
+    val uuidBytes: ByteBuffer = ByteBuffer.wrap(new Array[Byte](16))
```
No need for the `ByteBuffer` type annotation.

ijuma commented on core/src/main/scala/kafka/utils/CoreUtils.scala:
```diff
@@ -277,4 +279,18 @@ object CoreUtils extends Logging {
+    val uuidBytes: ByteBuffer = ByteBuffer.wrap(new Array[Byte](16))
+    uuidBytes.putLong(uuid.getMostSignificantBits)
+    uuidBytes.putLong(uuid.getLeastSignificantBits)
+    val base64EncodedUUID = DatatypeConverter.printBase64Binary(uuidBytes.array())
```
No need for `()` after `array`.

arrawatia committed:
Add tests for base64 encoding and clusterId generation. Fix some typos and errors.

ijuma commented on core/src/test/scala/unit/kafka/utils/UtilsTest.scala:
```diff
@@ -151,4 +154,25 @@ class UtilsTest extends JUnitSuite {
+    assertTrue(clusterIdPattern.matcher(clusterId1).matches())
+
+    // Test a UUID that has + or / characters in base64 encoding
+    // d418ec02-277e-4853-81e6-afe30259daec ->(base64)-> 1BjsAid+SFOB5q/jAlna7A==
```
Seems like we can assert the expected value too instead of having it in a comment?

arrawatia committed:
Fix upgrade test.

arrawatia committed:
Fix upgrade test again.

arrawatia committed:
Fixes for @ijuma 's review comments.

ijuma commented:
Can you please fix the PR description to remove "first draft"? The PR description becomes the commit message so that's why it's important to have it match what will be merged.

arrawatia committed:
Fixes for @ijuma 's review comments.

arrawatia commented:
Done. I updated https://issues.apache.org/jira/browse/KAFKA-4151 to track the docs for KIP-78 (The JIRA ticket was tracking systemtests previously)

arrawatia commented:
Update : all tests in upgrade_test.py pass.

[Branch Builder](http://jenkins.confluent.io/job/system-test-kafka-branch-builder/543/console)

```
================================================================================
SESSION REPORT (ALL TESTS)
session_id: 2016-09-15--001
run time:   33 minutes 30.015 seconds
tests run:  12
passed:     12
failed:     0
ignored:    0
================================================================================
test_id:    2016-09-15--001.kafkatest.tests.core.upgrade_test.TestUpgrade.test_upgrade.from_kafka_version=0.10.0.1.to_message_format_version=None.compression_types=.none
status:     PASS
run time:   2 minutes 41.042 seconds
--------------------------------------------------------------------------------
test_id:    2016-09-15--001.kafkatest.tests.core.upgrade_test.TestUpgrade.test_upgrade.from_kafka_version=0.9.0.1.to_message_format_version=None.compression_types=.none
status:     PASS
run time:   2 minutes 44.854 seconds
--------------------------------------------------------------------------------
test_id:    2016-09-15--001.kafkatest.tests.core.upgrade_test.TestUpgrade.test_upgrade.from_kafka_version=0.9.0.1.to_message_format_version=None.security_protocol=SASL_SSL.compression_types=.none.new_consumer=True
status:     PASS
run time:   4 minutes 22.439 seconds
--------------------------------------------------------------------------------
test_id:    2016-09-15--001.kafkatest.tests.core.upgrade_test.TestUpgrade.test_upgrade.from_kafka_version=0.9.0.1.to_message_format_version=None.compression_types=.snappy.new_consumer=True
status:     PASS
run time:   2 minutes 44.910 seconds
--------------------------------------------------------------------------------
test_id:    2016-09-15--001.kafkatest.tests.core.upgrade_test.TestUpgrade.test_upgrade.from_kafka_version=0.9.0.1.to_message_format_version=None.compression_types=.lz4
status:     PASS
run time:   2 minutes 36.475 seconds
--------------------------------------------------------------------------------
test_id:    2016-09-15--001.kafkatest.tests.core.upgrade_test.TestUpgrade.test_upgrade.from_kafka_version=0.9.0.1.to_message_format_version=None.compression_types=.lz4.new_consumer=True
status:     PASS
run time:   2 minutes 44.844 seconds
--------------------------------------------------------------------------------
test_id:    2016-09-15--001.kafkatest.tests.core.upgrade_test.TestUpgrade.test_upgrade.from_kafka_version=0.9.0.1.to_message_format_version=0.9.0.1.compression_types=.none
status:     PASS
run time:   2 minutes 40.207 seconds
--------------------------------------------------------------------------------
test_id:    2016-09-15--001.kafkatest.tests.core.upgrade_test.TestUpgrade.test_upgrade.from_kafka_version=0.9.0.1.to_message_format_version=0.9.0.1.compression_types=.snappy.new_consumer=True
status:     PASS
run time:   2 minutes 48.091 seconds
--------------------------------------------------------------------------------
test_id:    2016-09-15--001.kafkatest.tests.core.upgrade_test.TestUpgrade.test_upgrade.from_kafka_version=0.9.0.1.to_message_format_version=0.9.0.1.compression_types=.lz4
status:     PASS
run time:   2 minutes 31.521 seconds
--------------------------------------------------------------------------------
test_id:    2016-09-15--001.kafkatest.tests.core.upgrade_test.TestUpgrade.test_upgrade.from_kafka_version=0.9.0.1.to_message_format_version=0.9.0.1.compression_types=.lz4.new_consumer=True
status:     PASS
run time:   2 minutes 42.482 seconds
--------------------------------------------------------------------------------
test_id:    2016-09-15--001.kafkatest.tests.core.upgrade_test.TestUpgrade.test_upgrade.from_kafka_version=0.8.2.2.to_message_format_version=None.compression_types=.none
status:     PASS
run time:   2 minutes 31.080 seconds
--------------------------------------------------------------------------------
test_id:    2016-09-15--001.kafkatest.tests.core.upgrade_test.TestUpgrade.test_upgrade.from_kafka_version=0.8.2.2.to_message_format_version=None.compression_types=.snappy
status:     PASS
run time:   2 minutes 21.856 seconds
--------------------------------------------------------------------------------

```

arrawatia committed:
Merge branch 'trunk' of github.com:apache/kafka into kip-78

arrawatia committed:
Merge branch 'trunk' of github.com:apache/kafka into kip-78

ijuma committed:
Various clean-ups (#2)

arrawatia committed:
Merge branch 'trunk' of github.com:apache/kafka into kip-78

arrawatia committed:
Merge branch 'kip-78' of github.com:arrawatia/kafka into kip-78

arrawatia committed:
Clean up the state properly to fix flaky tests.

ijuma commented:
LGTM, thanks for the PR. :)
